### PR TITLE
kanidm: fix build for rust v1.75

### LINKED ISSUE #281018: Build failure: kanidm
### Steps To Reproduce

1. build kanidm

### Build log

```
https://hydra.nixos.org/build/245589915
```

### Notify maintainers
@erictapen @Flakebi @mweinelt @adamcstephens

### Metadata
```console
NixOS, 24.05 (Uakari), 24.05.20231227.f930306
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Kranzes**
## Description of changes

closes: #281018 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by adamcstephens — APPROVED** [review]
Result of `nixpkgs-review pr 281023` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kanidm</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 281023` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kanidm</li>
  </ul>
</details>

**Review by oddlama — APPROVED** [review]

File: pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ use kanidmd_web_ui_shared::utils::{do_alert_error, do_footer, window};
+diff --git a/unix_integration/nss_kanidm/src/lib.rs b/unix_integration/nss_kanidm/src/lib.rs
+index ef13192b9..27e3321a8 100644
+--- a/unix_integration/nss_kanidm/src/lib.rs
++++ b/unix_integration/nss_kanidm/src/lib.rs
+@@ -20,5 +20,4 @@ extern crate lazy_static;
+ #[cfg(target_family = "unix")]
+ mod implementation;
+ 
+-#[cfg(target_family = "unix")]
```
Comment (yaleman) [inline pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch:76]: Removing these lines effectively stops the module from doing *anything*! It's the whole implementation 😄

**Review by adamcstephens — CHANGES_REQUESTED** [review]
See @yaleman's comment.

Can you also move the source forward to the latest commit on the rc branch? It fixes the version which includes `-dev` in it.

https://github.com/kanidm/kanidm/commits/1.1.0-rc.15/

File: pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch (hunk 1 above)
Comment (Kranzes) [reply pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch:76]: I guess `cargo fix` just completely breaks it. Can you point me out to what needs to be done?

Comment (yaleman) [reply pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch:76]: You should just have to update to the head of the tag, as far as I can tell?

**Comment by Kranzes** [comment]
It seems that the modifications in the patch are already part of the master branch but not part of the rc branch which we use. For now I have made the patch myself but a better solution would be to have @yaleman backport https://github.com/kanidm/kanidm/commit/dbf476fe5ea2c120dff9a85e552be9f898c69ce7 into the branch we use.

On another note, the VM tests are broken and get stuck due to a missing path to a bash shell, can anyone take a look at that?

**Comment by adamcstephens** [comment]
This still fails for me:

```
  Compiling nss_kanidm v1.1.0-rc.15 (/build/source/unix_integration/nss_kanidm)
error[E0277]: a value of type `hashbrown::HashSet<(&SmartString<LazyCompact>, &value::IndexType)>` cannot be built from an iterator over elements of type `&(_, value::IndexType)`
   --> server/lib/src/filter.rs:581:42
    |
581 |         let idxmeta_ref = idxmeta.iter().collect();
    |                                          ^^^^^^^ value of type `hashbrown::HashSet<(&SmartString<LazyCompact>, &value::IndexType)>` cannot be built from `std::iter::Iterator<Item=&(_, value::IndexType)>`
    |
    = help: the trait `FromIterator<&(_, value::IndexType)>` is not implemented for `hashbrown::HashSet<(&SmartString<LazyCompact>, &value::IndexType)>`
    = help: the trait `FromIterator<(&SmartString<LazyCompact>, &value::IndexType)>` is implemented for `hashbrown::HashSet<(&SmartString<LazyCompact>, &value::IndexType)>`
    = help: for that trait implementation, expected `(&SmartString<LazyCompact>, &value::IndexType)`, found `&(_, value::IndexType)`
note: the method call chain might not have had the expected associated types
   --> server/lib/src/filter.rs:581:35
    |
565 |           let idxmeta = vec![
    |  _______________________-
566 | |             (Attribute::Uuid.into(), IndexType::Equality),
567 | |             (Attribute::Uuid.into(), IndexType::Presence),
568 | |             (Attribute::Name.into(), IndexType::Equality),
...   |
578 | |             (Attribute::DirectMemberOf.into(), IndexType::Presence),
579 | |         ];
    | |_________- this expression has type `Vec<(_, IndexType)>`
580 |
581 |           let idxmeta_ref = idxmeta.iter().collect();
    |                                     ^^^^^^ `Iterator::Item` is `&(_, IndexType)` here
note: required by a bound in `std::iter::Iterator::collect`
   --> /build/rustc-1.75.0-src/library/core/src/iter/traits/iterator.rs:2050:5
```

**Comment by Kranzes** [comment]
@adamcstephens Can you try again?

**Review by adamcstephens — APPROVED** [review]
```
─❯ /nix/store/c5v5c642p46sdfxhr767dj0j77yprsis-kanidm-1.1.0-rc.15/bin/kanidm version
kanidm 1.1.0-rc.15
2024-01-15T21:12:45.736526Z DEBUG kanidm: Using 24 worker threads
```

Result of `nixpkgs-review pr 281023` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kanidm</li>
  </ul>
</details>

-----

Result of `nixpkgs-review pr 281023` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>kanidm</li>
  </ul>
</details>

File: pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch (hunk 1 above)
Comment (yaleman) [reply pkgs/servers/kanidm/0001-fix-warnings-for-rust-v1.75.patch:76]: Urgh I'm dumb ignore all this 🙄

**Review by oddlama — APPROVED** [review]
Can confirm, builds on x86_64-linux and works.

**Comment by matthew-salerno** [comment]
Should this be backported to stable? I just ran into it today.

**Comment by adamcstephens** [comment]
Yes. I guess I'd assumed the rust changes wouldn't have gone to stable, but...the kanidm build is indeed failing there.

**Comment by Kranzes** [comment]
> Yes. I guess I'd assumed the rust changes wouldn't have gone to stable, but...the kanidm build is indeed failing there.

Is it?

**Comment by ttamttamS** [comment]
It might not be, just realized the machine I got the failure for was on unstable. Will investigate later today. Might be user error

**Comment by adamcstephens** [comment]
I tested the build locally on `nixos-23.11` and it failed to build. I’ve no idea what hydra-check is reporting but that date is from November.

**Comment by Kranzes** [comment]
Ok feel free to merge the backport.

**Comment by adamcstephens** [comment]
I guess I'm actually confused, because I just rechecked and that cached build is correct. So it's not required, as hydra-check is correctly reporting, but I don't think the backport will hurt. If anything it'll fix the version label in stable.

**Comment by matthew-salerno** [comment]
I'm confused too, the particular machine I had the issue on was actually on unstable. I forgot I changed it to unstable.

**Pull request merged**